In [2]:
import pandas as pd

In [3]:
df = pd.read_csv(r"C:\Users\mw3595\Documents\data vis project\311_Service_Requests_2024_2025.csv",on_bad_lines="skip")

C:\Users\mw3595\AppData\Local\Temp\ipykernel_31644\1061516281.py:1: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(r"C:\Users\mw3595\Documents\data vis project\311_Service_Requests_2024_2025.csv",on_bad_lines="skip")


https://data.cityofnewyork.us/Social-Services/311-Service-Requests-from-2020-to-Present/erm2-nwe9/explore/query/SELECT%0A%20%20%60unique_key%60%2C%0A%20%20%60created_date%60%2C%0A%20%20%60closed_date%60%2C%0A%20%20%60agency%60%2C%0A%20%20%60agency_name%60%2C%0A%20%20%60complaint_type%60%2C%0A%20%20%60descriptor%60%2C%0A%20%20%60location_type%60%2C%0A%20%20%60incident_zip%60%2C%0A%20%20%60incident_address%60%2C%0A%20%20%60street_name%60%2C%0A%20%20%60cross_street_1%60%2C%0A%20%20%60cross_street_2%60%2C%0A%20%20%60intersection_street_1%60%2C%0A%20%20%60intersection_street_2%60%2C%0A%20%20%60address_type%60%2C%0A%20%20%60city%60%2C%0A%20%20%60landmark%60%2C%0A%20%20%60facility_type%60%2C%0A%20%20%60status%60%2C%0A%20%20%60due_date%60%2C%0A%20%20%60resolution_description%60%2C%0A%20%20%60resolution_action_updated_date%60%2C%0A%20%20%60community_board%60%2C%0A%20%20%60bbl%60%2C%0A%20%20%60borough%60%2C%0A%20%20%60x_coordinate_state_plane%60%2C%0A%20%20%60y_coordinate_state_plane%60%2C%0A%20%20%60open_data_channel_type%60%2C%0A%20%20%60park_borough%60%2C%0A%20%20%60latitude%60%2C%0A%20%20%60longitude%60%2C%0A%20%20%60location%60%0AWHERE%0A%20%20%60created_date%60%0A%20%20%20%20BETWEEN%20%222024-01-01T16%3A23%3A16%22%20%3A%3A%20floating_timestamp%0A%20%20%20%20AND%20%222025-12-31T16%3A23%3A16%22%20%3A%3A%20floating_timestamp%0A%20%20AND%20caseless_ne%28%60agency%60%2C%20%22HPD%22%29%0AORDER%20BY%20%60created_date%60%20DESC%20NULL%20FIRST/page/filter


#### The downloaded data contains only 911 complaints for 2024 and 2025. File excludes HPD (New York City Department of Housing Preservation and Development) because it is usually concerned only with building conditions and is not relevant to our project conceptually.


**Data also does not include those columns:**
    "vehicle_type",
    "taxi_company_borough",
    "taxi_pick_up_location",
    "bridge_highway_name",
    "bridge_highway_direction",
    "road_ramp",
    "bridge_highway_segment",
    "due_date", 
    "resolution_description",
    "resolution_action_updated_date",
    "landmark",
    "facility_type" 

In [4]:
df.head()

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Problem (formerly Complaint Type),Problem Detail (formerly Descriptor),Location Type,Incident Zip,Incident Address,...,Community Board,BBL,Borough,X Coordinate (State Plane),Y Coordinate (State Plane),Open Data Channel Type,Park Borough,Latitude,Longitude,Location
0,67347796,12/31/2025 04:23:00 PM,01/01/2026 01:00:00 AM,DEP,Department of Environmental Protection,Water System,Dirty Water (WE),NaN,10031.0,626 RIVERSIDE DRIVE,...,09 MANHATTAN,1.020870e+09,MANHATTAN,"996,862","239,357",ONLINE,MANHATTAN,40.823646,-73.954431,POINT (-73.954430909076 40.82364624134)
1,67346303,12/31/2025 04:22:33 PM,01/02/2026 01:35:00 PM,DOT,Department of Transportation,Street Condition,Pothole,NaN,10461.0,INDUSTRIAL STREET,...,11 BRONX,NaN,BRONX,NaN,NaN,UNKNOWN,BRONX,NaN,NaN,NaN
2,67347638,12/31/2025 04:22:31 PM,01/02/2026 08:31:35 AM,DOT,Department of Transportation,Street Sign - Missing,Other/Unknown,Street,10314.0,4113 VICTORY BOULEVARD,...,02 STATEN ISLAND,5.027670e+09,STATEN ISLAND,"931,214","154,849",MOBILE,STATEN ISLAND,40.591542,-74.190961,POINT (-74.190961278787 40.591542246598)
3,67346505,12/31/2025 04:22:09 PM,01/07/2026 08:54:28 PM,DSNY,Department of Sanitation,Dirty Condition,Trash,Street,10026.0,ST NICHOLAS AVENUE,...,10 MANHATTAN,NaN,MANHATTAN,"997,313","231,960",PHONE,MANHATTAN,40.803343,-73.952816,POINT (-73.952815773413 40.803342887744)
4,67351565,12/31/2025 04:21:58 PM,12/31/2025 04:37:20 PM,NYPD,New York City Police Department,Noise - Vehicle,Car/Truck Horn,Street/Sidewalk,11366.0,78-15 PARSONS BOULEVARD,...,08 QUEENS,4.068290e+09,QUEENS,"1,037,000","202,363",ONLINE,QUEENS,40.721959,-73.809697,POINT (-73.809696824262 40.721959131993)


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5597361 entries, 0 to 5597360
Data columns (total 33 columns):
 #   Column                                Dtype  
---  ------                                -----  
 0   Unique Key                            int64  
 1   Created Date                          object 
 2   Closed Date                           object 
 3   Agency                                object 
 4   Agency Name                           object 
 5   Problem (formerly Complaint Type)     object 
 6   Problem Detail (formerly Descriptor)  object 
 7   Location Type                         object 
 8   Incident Zip                          object 
 9   Incident Address                      object 
 10  Street Name                           object 
 11  Cross Street 1                        object 
 12  Cross Street 2                        object 
 13  Intersection Street 1                 object 
 14  Intersection Street 2                 object 
 15  Address Type   

In [7]:
df["Created Date"] = pd.to_datetime(
    df["Created Date"],
    format="%m/%d/%Y %I:%M:%S %p",
    errors="coerce"
)

df["Closed Date"] = pd.to_datetime(
    df["Closed Date"],
    format="%m/%d/%Y %I:%M:%S %p",
    errors="coerce"
)

In [8]:
df["Complaint Year"] = df["Created Date"].dt.year

In [9]:
df["Complaint Month"] = df["Created Date"].dt.month

In [10]:
print(df["Complaint Year"].unique())
df["Complaint Month"].unique()

[2025 2024]


array([12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1])

In [12]:
df["Month Name"] = pd.to_datetime(df["Complaint Month"], format="%m").dt.month_name()

In [13]:
df["Month Name"].unique()

array(['December', 'November', 'October', 'September', 'August', 'July',
       'June', 'May', 'April', 'March', 'February', 'January'],
      dtype=object)

In [14]:
df["Agency"].unique()

array(['DEP', 'DOT', 'DSNY', 'NYPD', 'DOB', 'DOHMH', 'OOS', 'TLC', 'EDC',
       'DCWP', 'DPR', 'DHS', 'DOE', 'OTI', '3-1-1'], dtype=object)

In [44]:
#df = df[df["Agency"] != "HPD"]
#df["Agency"].unique()

In [26]:
#df["Problem (formerly Complaint Type)"].unique()

In [15]:
#for the ease of use - we will rename the columns
df = df.rename(columns={
    "Problem (formerly Complaint Type)": "Complaint",
    "Problem Detail (formerly Descriptor)": "Complaint Detail"
})

In [25]:
df = df.rename(columns={
    'X Coordinate (State Plane)': "X Coordinate",
    'Y Coordinate (State Plane)': "Y Coordinate"
})

In [16]:
grouping_complaints = (
    df.groupby(["Agency", "Complaint","Complaint Detail"])
      .size()
      .reset_index(name="count")
)

In [17]:
grouping_complaints

,Agency,Complaint,Complaint Detail,count
0,3-1-1,Noise - Commercial,Loud Music/Party,1
1,DCWP,Consumer Complaint,Airline,8
2,DCWP,Consumer Complaint,Arcade,9
3,DCWP,Consumer Complaint,Art Supply Store,11
4,DCWP,Consumer Complaint,Auction House or Auctioneer,18
...,...,...,...,...
1142,TLC,Taxi Licensee Complaint,Owner/Broker/Agent Complaint,4
1143,TLC,Taxi Licensee Complaint,Vehicle Report,1
1144,TLC,Taxi Report,Driver Report - Non Passenger,314
1145,TLC,Taxi Report,Driver Report - Passenger,1103


In [18]:
#if you want your own table :
# grouping_complaints.to_excel("complaints_by_agency_problem.xlsx", index=False)

In [36]:
df.isnull().sum()

Unique Key                              0
Created Date                            0
Closed Date                        136499
Agency                                  0
Agency Name                             0
Complaint                               0
Complaint Detail                   196265
Location Type                      845736
Incident Zip                        64029
Incident Address                   242567
Street Name                        242798
Cross Street 1                     501565
Cross Street 2                     500465
Intersection Street 1              782735
Intersection Street 2              779527
Address Type                        28434
City                               317112
Landmark                          1267524
Facility Type                     5571498
Status                                  0
Due Date                          5561175
Resolution Description             138109
Resolution Action Updated Date      56759
Community Board                   

We have 5,597,360 rows and for:
- 102,672 we do not have Location
- 102,639 we do not have X Coordinate
- 102,458 we do not have Y Coordinate

In [27]:
df[(df["Location"].notna()) & (df["X Coordinate"].isna())]

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


In [40]:
rows_with_location_missing_x = df[
    df["Location"].notna() & df["X Coordinate"].isna()
]

print("Number of rows:", len(rows_with_location_missing_x))  # should print 33
rows_with_location_missing_x.head()

Number of rows: 0


,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


In [42]:
df[(df["Location"].notna()) & (df["Y Coordinate"].isna())]

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name


In [31]:
df[["Location","X Coordinate"]].isna().value_counts()

Location  X Coordinate
False     False           5494689
True      True             102639
          False                33
Name: count, dtype: int64

5,494,689 rows → both Location and X Coordinate (State Plane) exist ✅

102,639 rows → Location is missing, X Coordinate is missing ❌

33 rows → X Coordinate exists but Location is missing ⚠

In [41]:
rows_with_location_missing_x = df[
    df["X Coordinate"].notna() & df["Location"].isna()
]

len(rows_with_location_missing_x)  # should print 33
rows_with_location_missing_x

,Unique Key,Created Date,Closed Date,Agency,Agency Name,Complaint,Complaint Detail,Location Type,Incident Zip,Incident Address,...,X Coordinate,Y Coordinate,Open Data Channel Type,Park Borough,Latitude,Longitude,Location,Complaint Year,Complaint Month,Month Name
83653,67224277,2025-12-17 14:44:57,2025-12-18 14:15:05,DOT,Department of Transportation,Street Condition,Blocked - Construction,Street,11201.0,CLARK STREET,...,"985,392",NaN,PHONE,BROOKLYN,NaN,NaN,NaN,2025,12,December
176201,67082296,2025-12-06 17:31:33,2025-12-09 08:44:14,DOT,Department of Transportation,Broken Parking Meter,Out of Order,Street,10075.0,EAST 78 STREET,...,"994,680",NaN,ONLINE,MANHATTAN,NaN,NaN,NaN,2025,12,December
298168,66901503,2025-11-20 16:20:18,NaT,DOE,Department of Education,School Maintenance,Other School Condition,School,NaN,62-10,...,"1,025,093",NaN,ONLINE,QUEENS,NaN,NaN,NaN,2025,11,November
346429,66829282,2025-11-13 19:02:33,2025-11-17 09:25:16,DOT,Department of Transportation,Street Condition,"Rough, Pitted or Cracked Roads",Street,10452.0,RIVER AVENUE,...,"1,006,269",NaN,PHONE,BRONX,NaN,NaN,NaN,2025,11,November
396605,66759483,2025-11-07 06:53:45,2025-11-10 12:57:50,DOT,Department of Transportation,Street Condition,Cave-in,Street,11208.0,RIDGEWOOD AVENUE,...,"1,017,810",NaN,UNKNOWN,BROOKLYN,NaN,NaN,NaN,2025,11,November
466264,66651110,2025-10-29 16:00:37,2025-10-31 11:33:00,DCWP,Department of Consumer and Worker Protection,Consumer Complaint,Ticket Seller,Business,NaN,1236,...,"1,006,018",NaN,ONLINE,BRONX,NaN,NaN,NaN,2025,10,October
1174794,65783225,2025-08-07 06:55:29,2025-08-07 10:06:57,DOT,Department of Transportation,Broken Parking Meter,No Receipt,Street,10001.0,WEST 28 STREET,...,"986,393",NaN,PHONE,MANHATTAN,NaN,NaN,NaN,2025,8,August
2166450,64619746,2025-04-11 16:22:49,2025-04-17 11:45:32,DOT,Department of Transportation,Highway Condition,Pothole - Highway,Highway,NaN,NaN,...,"1,000,115",NaN,PHONE,MANHATTAN,NaN,NaN,NaN,2025,4,April
2747820,63777089,2025-01-15 09:13:05,2025-02-11 07:31:48,DOT,Department of Transportation,Broken Parking Meter,No Receipt,Street,10462.0,WESTCHESTER AVENUE,...,"1,023,085",NaN,PHONE,BRONX,NaN,NaN,NaN,2025,1,January
2801240,63704297,2025-01-09 10:34:00,2025-02-19 13:31:04,DOE,Department of Education,School Maintenance,Heating Problem,School,NaN,750,...,"1,030,803",NaN,ONLINE,BRONX,NaN,NaN,NaN,2025,1,January


We are removing all 33 where we do not have location and both Y and X coordinates

In [43]:
# Keep only rows where Location is not null
df = df[df["Location"].notna()]

# Quick check
df["Location"].isna().sum()  # should be 0

0